### PR TITLE
Add registry functionality to Symphony Electron

### PR DESCRIPTION
**Pull request by rbarros-daitan**

### CODE REVIEW

File: js/getRegistry.js [hunk 1]
```diff
@@ -0,0 +1,55 @@
+'use strict';
+
+const symphonyRegistry = '\\Software\\Symphony\\Symphony\\';
+
+var Registry = require('winreg')
```
Comment (lneir) [inline js/getRegistry.js]: how are you requiring this?  i don't see in package.json?

File: js/getRegistry.js [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+    
+/**
+ * reads Windows Registry key. This Registry holds the Symphony registry keys 
+ * that are intended to be used as global (or default) value for all users 
+ * running this app.
+ *
+ * @return {Object} Windows registry key
+ */
+function getRegistry(name) {
+    return new Promise(function(resolve, reject) {
```
Comment (lneir) [inline js/getRegistry.js]: doesn't this only work on windows?  this code base also supports mac.

File: js/getRegistry.js [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+
+        //Try to get registry on HKEY_CURRENT_USER    
+        symphonyRegistryHKCU.get( name, function( err1, reg1 ) {
+            if ( !err1 && reg1!=null && reg1.value) {
```
Comment (lneir) [inline js/getRegistry.js]: should use !==

File: js/getRegistry.js [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+                    if ( !err2 && reg2!=null && reg2.value) {
+                        resolve(reg2.value);
+                    } else{
+                        //Try to get registry on HKEY_LOCAL_MACHINE in case 32bit app installed on 64bit system.
+                        //winreg does not merge keys as normally windows does.
+                        symphonyRegistryHKLM6432.get( name, function( err3, reg3 ) {
+                            if ( !err3 && reg3!=null && reg3.value) {
+                                resolve(reg3.value);
+                            } else{
+                                reject();
```
Comment (lneir) [inline js/getRegistry.js]: reject should provide error msg.

File: js/getRegistry.js [hunk 5]
```diff
@@ -0,0 +1,55 @@
...
+    return new Promise(function(resolve, reject) {
+
+        //Try to get registry on HKEY_CURRENT_USER    
+        symphonyRegistryHKCU.get( name, function( err1, reg1 ) {
+            if ( !err1 && reg1!=null && reg1.value) {
+                resolve(reg1.value);
+            } else{
+                //Try to get registry on HKEY_LOCAL_MACHINE                    
+                symphonyRegistryHKLM.get( name, function( err2, reg2 ) {
+                    if ( !err2 && reg2!=null && reg2.value) {
```
Comment (lneir) [inline js/getRegistry.js]: ditto above !==

File: js/getRegistry.js [hunk 6]
```diff
@@ -0,0 +1,55 @@
...
+    hive: Registry.HKCU,
+    key:  symphonyRegistry
+})
+
+, symphonyRegistryHKLM = new Registry({
+    key:  symphonyRegistry
+})
+
+, symphonyRegistryHKLM6432 = new Registry({
+    key:  symphonyRegistry.replace("\\Software","\\Software\\WOW6432Node")
```
Comment (lneir) [inline js/getRegistry.js]: use single quotes.

File: js/getRegistry.js (hunk 1 above)
Comment (lneir) [inline js/getRegistry.js]: missing semicolon

File: js/main.js [hunk 7]
```diff
@@ -29,24 +30,32 @@ app.on('ready', getUrlAndOpenMainWindow);
...
-        // add https protocol if none found.
-        let parsedUrl = nodeURL.parse(config.url);
-        if (!parsedUrl.protocol) {
-            protocol = 'https';
-        }
-        var url = nodeURL.format({
-            protocol: protocol,
-            slahes: true,
-            pathname: parsedUrl.href
+        getRegistry("PodUrl").then(function(podUrl){
```
Comment (lneir) [inline js/main.js]: single quotes.

Comment (lneir) [inline js/main.js]: you should chain the promises.

File: js/getRegistry.js (hunk 1 above)
Comment (rbarros-daitan) [inline js/getRegistry.js]: @Added on package.json. Sorry about that.

File: js/getRegistry.js (hunk 2 above)
Comment (rbarros-daitan) [inline js/getRegistry.js]: Added isMac to check for Mac OS.

File: js/getRegistry.js (hunk 3 above)
Comment (rbarros-daitan) [inline js/getRegistry.js]: ok.

File: js/getRegistry.js (hunk 4 above)
Comment (rbarros-daitan) [inline js/getRegistry.js]: ok.

File: js/getRegistry.js (hunk 5 above)
Comment (rbarros-daitan) [inline js/getRegistry.js]: ok.

File: js/getRegistry.js (hunk 6 above)
Comment (rbarros-daitan) [inline js/getRegistry.js]: ok.

File: js/main.js (hunk 7 above)
Comment (rbarros-daitan) [inline js/main.js]: ok.

File: js/getRegistry.js (hunk 1 above)
Comment (rbarros-daitan) [inline js/getRegistry.js]: This is another typing to define those var but i´ve changed to the regular typing to avoid confusion.

File: js/getRegistry.js [hunk 8]
```diff
@@ -0,0 +1,55 @@
+'use strict';
```
Comment (lneir) [inline js/getRegistry.js]: also let's move this into utils directory.  seems like a good spot for it.

**Comment by lneir** [comment]
also make sure you run this through linter: npm run lint

File: js/getRegistry.js (hunk 8 above)
Comment (rbarros-daitan) [inline js/getRegistry.js]: file relocated.

**Comment by rbarros-daitan** [comment]
@lneir , hi. I´ve applied the suggested changes. Can you please review it again? Thanks.

File: js/getConfig.js [hunk 9]
```diff
@@ -41,13 +39,14 @@ function getConfig() {
                 try {
                     // data is the contents of the text file we just read
                     let config = JSON.parse(data);
-                    resolve(config);
+                    resolve(config.url);
```
Comment (lneir) [inline js/getConfig.js]: should be returning entire config here.

File: js/main.js [hunk 10]
```diff
@@ -28,25 +29,31 @@ const windowMgr = require('./windowMgr.js');
...
-        let parsedUrl = nodeURL.parse(config.url);
-        if (!parsedUrl.protocol) {
-            protocol = 'https';
-        }
-        var url = nodeURL.format({
-            protocol: protocol,
-            slahes: true,
-            pathname: parsedUrl.href
+    getRegistry('PodUrl')
+    .then(createWin).catch(function (){
```
Comment (lneir) [inline js/main.js]: this promise chain is bit hard to follow...
think it might be better to move getRegistry call inside of getConfig and replace returned url there with one from registry (if window and presents).

also make sure unit tests still run: npm run test

File: js/getConfig.js (hunk 9 above)
Comment (rbarros-daitan) [inline js/getConfig.js]: ok. I had to change createWin to accept both string & object parameter.

File: js/getConfig.js [hunk 11]
```diff
@@ -38,16 +36,24 @@ function getConfig() {
             if (err) {
                 reject('cannot open config file: ' + configPath + ', error: ' + err);
             } else {
+                let config = '';
```
Comment (lneir) [inline js/getConfig.js]: let's make config an object: let config = {};

File: js/main.js [hunk 12]
```diff
@@ -28,25 +28,36 @@ const windowMgr = require('./windowMgr.js');
...
-    }).catch(function(err) {
+    getConfig()
+    .then(createWin).catch(function (err){
         let title = 'Error loading configuration';
-        electron.dialog.showErrorBox(title, title + ': ' + err);
+        electron.dialog.showErrorBox(title, title + ': ' + err);            
     });
 }
 
+function createWin(podurl){
```
Comment (lneir) [inline js/main.js]: the signature here should be config: createWin(config)

the idea is that getConfig will return object, for now it just return object with url but in future there could be other things here.

Comment (rbarros-daitan) [inline js/main.js]: ok

File: js/getConfig.js (hunk 11 above)
Comment (rbarros-daitan) [inline js/getConfig.js]: ok.

**Comment by rbarros-daitan** [comment]
@lneir , latest comments were applied. Could you please check? Thanks.

File: .eslintignore [hunk 13]
```diff
@@ -6,3 +6,4 @@ installer
 node_modules
 js/preload/_*.js
 tests/**
+js/notify/*.html
```
Comment (lneir) [inline .eslintignore]: why this?

Comment (rbarros-daitan) [inline .eslintignore]: lint complains about this html file that was added by you.

Comment (lneir) [inline .eslintignore]: i'm not seeing lint errors.  
npm run lint looks fine.
also .html shouldn't be run through linter, just .js files.
